### PR TITLE
Add transport version to cluster state diff

### PR DESCRIPTION
**Pull request by thecoop**
Followon from #94157, add transport version to cluster state diff.

Also ensure that everywhere `ClusterState.nodes` is used (not in a test), `transportVersions` is updated too.

Note that any inferred transport versions will stay in the cluster state until that node is restarted. #95148 adds a cluster state listener to eagerly update any inferred transport versions to the correct versions.

That PR is not actually needed until a transport version is used that is not 8.8.0 (the inferred version is currently the correct version, as 8.8.0 is the most recent version, so no fixups are actually required), so #95148 is actually only needed for 8.8.1/8.9.0

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 1]
```diff
@@ -756,6 +769,11 @@ public Builder putTransportVersion(String node, TransportVersion version) {
             return this;
         }
 
+        public Builder removeTransportVersion(String type) {
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/ClusterState.java:772]: NIT: I believe the key here is node?

**Comment by thecoop** [comment]
The bit I'm not sure about is - if cluster state is always passed around using diffs, is it possible for a node's transport version to get 'stuck' and always inferred as the minimum version?

I'm pretty sure its fine, as _at some point_ every node will restart/rejoin the cluster when the master is a >=8.8.0 node, which will then set the transport version correctly in the master state, which will then get propagated to every other node over time.

The situation I'm concerned about is:
1. Master is <8.8.0, no transport version info
2. After upgrade, all nodes are >=8.8.0, but the cluster state is still that created by a <8.8.0 node, this will record the transport version of some nodes as 8.8.0 when they might be a higher version.
3. This persists over node restarts...somehow?

The worst case is that the minimum transport version is lower than it could be - but that's ok, as long as _at some point_ it moves forward from 8.8.0.

**Comment by DaveCTurner** [comment]
I think that concern is independent of the question of whether we send the state using diffs or not?

**Comment by thecoop** [comment]
It's more of a concern here, as in the state diff serialization there's now cases where the stream transport version is >=8.8.0 but the transportversions info is null, and then it'll be inferred (as you can't extract the diff information until it's been applied)

**Comment by DaveCTurner** [comment]
Are we allowing the cluster state to be different between the older and newer nodes by having the newer nodes infer the transport versions if they're missing? If so, that sounds like it's going to be wrong even without diffs?

**Comment by thecoop** [comment]
It doesn't particularly matter if its _wrong_, as long as it is updated _at some point in the future_ (when that particular node is restarted when the master is >=8.8.0) to be correct. The issue is if there's some case where the cluster state is kept in the new master with a null transportversion in the cluster diff it uses, and so always uses the inferred version even when that node is restarted. I don't believe this will happen, as I don't think the diff is persisted anywhere, but if someone in the distributed team could confirm this (with this new code) that would be great.

**Comment by DaveCTurner** [comment]
When a node is restarted it will leave the cluster and then join again from scratch. But is it not a concern that this data might be wrong for some arbitrary period of time before that happens? I'm uncomfortable with that. We had that in the past with node roles and it was awful from a supportability point of view.

**Comment by thecoop** [comment]
The only way this will be wrong is giving a lower min compat version than it otherwise could - which is ok. And we can't really do anything else, we need to infer the transport version when none is provided, and because there's no mapping between version and transport version we can only infer it as the earliest possible version to stop the version going backwards

**Comment by thecoop** [comment]
We could always infer a transport version specific to this case (if the first 'real' transport version is 10,000,000, use 999,999) to make it clear it's inferred - that will make it clearer what's going on if this does ever need to be debugged.

**Comment by DaveCTurner** [comment]
I would like that.

Moreover if we have such a marker then we can fix up the inferred versions eagerly once all the older nodes leave the cluster.

**Comment by thecoop** [comment]
OK, that would need to be implemented once we move to separate transport version numbers. I'll make a note

**Comment by DaveCTurner** [comment]
Not sure I follow - could we not do this in 8.8 by retrieving the transport version using the nodes-info action?

**Comment by DaveCTurner** [comment]
Ok got it

**Comment by thecoop** [comment]
Following on from discussions around the negative effects of keeping inferred versions around long-term, I've started a [separate PR](https://github.com/elastic/elasticsearch/pull/95148) to [fix up any inferred transport versions](https://github.com/thecoop/elasticsearch/blob/transportversion-inferredversion/server/src/main/java/org/elasticsearch/cluster/service/TransportVersionsFixupListener.java) once the whole cluster understands transport versions in cluster state.

Note that this can't actually be used until we have a value TransportVersion constant we can use to represent an inferred version, which can't be done until we've bumped to the first non-Version TransportVersion, which requires this PR to go in first.

So I suggest reviewing this PR on the understanding that the approach in #95148 (once finished & tested) closes the gap, and #95148 can be applied at the same time as the change to use the first separate TransportVersion

**Comment by thecoop** [comment]
As an alternative, I've pushed a commit that uses `Optional.empty` to indicate an inferred transport version - this can then be used to explicitly detect when an inferred version is present

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by thecoop** [comment]
Reverted PR to state as of [1e2c2c8](https://github.com/elastic/elasticsearch/pull/94870/commits/1e2c2c8928c668030d2113c75840122477e9c6dc)

**Review by DaveCTurner — APPROVED** [review]
LGTM2, on the grounds that we can tell if the state contains inferred versions because it would have nodes with a node version ≥8.8.1 but a tracked transport version ==8.8.0. At least if we infer version 8.8.0 for 8.8.0 nodes then that's correct, so no fixup is needed in this case.

**Pull request merged**